src/main/java/org/opentripplanner/street/model/edge/OsmEdge.java

```diff
@@ -0,0 +1,107 @@
+package org.opentripplanner.street.model.edge;
+
+import org.locationtech.jts.geom.Coordinate;
+import org.locationtech.jts.geom.LineString;
+import org.opentripplanner.framework.geometry.DirectionUtils;
+import org.opentripplanner.street.model.TurnRestriction;
+import org.opentripplanner.street.model.vertex.Vertex;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Abstract base class for edges derived from OSM data.
+ */
+public abstract class OsmEdge extends Edge {
```

Comment:
This should have different name and description but unfortunately I have no idea what.

Reply:
Adding classes in a class hierarchy that "only" provide some common code is usually a bad idea, it lead to fragmented code. An in this case also extra memory usage. The only logic I can see is in the constructor, that could be extracted and shared in other ways. Can the many setters and the lack of polymorphic behaviour be replaced with a builder instead?

Reply:
I agree that these setters are ugly but the problem here is that these setters are making the code _more_ polymorphic, not less. If we removed these, then I have to put `instanceOf` in many places to figure out if that edge handles them or not. These setters are not only used to construct the object but also to modify the edges later on in the build process.

I also want to get rid of them but that would be a larger refactoring. We have two options here: 

- Always use builders until the last possible moment in the build process and then build the graph at the very end.
- Make edges more immutable but allow it (perhaps via a builder) to replace them in a graph with a version with updated values.

All in all, I think that this is a topic for a dev meeting when Thomas is back from holiday.

Reply:
I will be there today, if not I will try to join next Thursday. I think the second option is the best choice. To do that you can do the following (psudo code):
"""
abstract class AbstractEdge<T> {
  /** Default implementation does nothing!!! */
  T withX(x : X) : T { return this; }
}

class ElevatorEdge extends AbstractEdge<ElevatorEdge> {

  ElevatorEdge withX(x : T) : ElevatorEdge { 
    var newEdge = new ElevatorEdge(..., x);
    replace_ this_with_newEdge();
    return newEdge; 
  }
}
"""